go-controller/pkg/node/vrfmanager/vrf_manager.go

```diff
@@ -0,0 +1,321 @@
+			}
+			mustRecreate = true
+		}
+	}
+	// Create VRF device if it doesn't exist or if it's needed to be recreated.
+	if util.GetNetLinkOps().IsLinkNotFoundError(err) || mustRecreate {
+		vrfLink = &netlink.Vrf{
+			LinkAttrs: netlink.LinkAttrs{Name: vrf.name},
+			Table:     vrf.table,
+		}
+		if err = util.GetNetLinkOps().LinkAdd(vrfLink); err != nil {
+			return fmt.Errorf("failed to create vrf device %s, err: %v", vrf.name, err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to retrieve existing vrf device %s, err: %v", vrf.name, err)
+	}
+	vrfLink, err = util.GetNetLinkOps().LinkByName(vrf.name)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve vrf device %s, err: %v", vrf.name, err)
+	}
+	if vrfLink.Attrs().OperState != netlink.OperUp {
+		if err = util.GetNetLinkOps().LinkSetUp(vrfLink); err != nil {
+			return fmt.Errorf("failed to get vrf device %s operationally up, err: %v", vrf.name, err)
+		}
+	}
+	existingEnslave, err := getExistingEnslaveInterfaceForVRF(vrfLink)
+	if err != nil {
+		return err
+	}
+	if existingEnslave == vrf.enslaveInterface {
+		return nil
+	}
+	if existingEnslave != "" {
+		err = removeInterfaceFromVRF(vrf.name, existingEnslave)
+		if err != nil {
+			return fmt.Errorf("failed to remove interface %s from vrf device: %s, err: %w", existingEnslave, vrf.name, err)
+		}
+	}
+	if vrf.enslaveInterface != "" {
+		err = enslaveInterfaceToVRF(vrf.name, vrf.enslaveInterface)
+		if err != nil {
+			return fmt.Errorf("failed to enslave interface %s into vrf device: %s, err: %w", vrf.enslaveInterface, vrf.name, err)
+		}
+	}
+	return nil
+}
+
+// AddVRF adds a VRF device into the node.
+func (vrfm *Controller) AddVRF(name, enslaveInterface string, table uint32) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+
+	vrfDev, ok := vrfm.vrfs[name]
+	if ok {
+		klog.V(5).Infof("VRF Manager: vrf %s already found in the cache", name)
+		if vrfDev.enslaveInterface != enslaveInterface {
+			return fmt.Errorf("VRF Manager: enslave interface mismatch for vrf device %s", name)
+		}
+		if vrfDev.table != table {
+			return fmt.Errorf("VRF Manager: table id mismatch for vrf device %s", name)
+		}
+	} else {
+		vrfDev = vrf{name, table, enslaveInterface, false}
+		vrfm.vrfs[name] = vrfDev
+	}
+	return vrfm.sync(vrfDev)
+}
+
+// Repair deletes stale VRF device(s) on the host. This helps remove
+// device(s) for which DeleteVRF is never invoked.
+// Assumptions: 1) The validVRFs list must contain device for which AddVRF
+// is already invoked. 2) The device name(s) in validVRFs are suffixed
+// with -vrf.
+func (vrfm *Controller) Repair(validVRFs sets.Set[string]) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+
+	links, err := util.GetNetLinkOps().LinkList()
+	if err != nil {
+		return fmt.Errorf("failed to list links on the node, err: %v", err)
+	}
+
+	for _, link := range links {
+		name := link.Attrs().Name
+		// Skip if the link is not a vrf type or name is not suffixed with -vrf.
```

Comment:
I wonder if -vrf isnt specific enough to OVNK. If a user had a vrf name with -vrf on the end of it, we could accidentally remove it. Maybe "udn-vrf" or something?

Reply:
ok, adding `-udn-vrf` as the suffix now. so now vrf device name would be like `mp<net-id>-udn-vrf`, net-id can accommodate only 5 digits, is this ok, i'm not sure ? 
or shall we keep as it is `mp<net-id>-vrf `and check for both prefix `mp` and suffix -`vrf` for repair ?